security/pkg/nodeagent/sds/sdsservice.go

```diff
@@ -18,576 +18,167 @@ package sds
-	if err != nil {
-		sdsServiceLog.Errorf("Failed to get secret for proxy %q from secret cache: %v", connID, err)
-		return nil, err
-	}
-
-	// Output the key and cert to a directory, if some applications need to read them from local file system.
-	if err = nodeagentutil.OutputKeyCertToDir(s.outputKeyCertToDir, secret.PrivateKey,
-		secret.CertificateChain, secret.RootCert); err != nil {
-		sdsServiceLog.Errorf("(%v) error when output the key and cert: %v",
-			connID, err)
-		return nil, err
-	}
-	return sdsDiscoveryResponse(secret, resourceName, discReq.TypeUrl)
+	return ret
 }
 
-func (s *sdsservice) getToken() (string, error) {
-	token := ""
-	if s.credFetcher != nil {
-		t, err := s.credFetcher.GetPlatformCredential()
+func (s *sdsservice) generate(resourceNames []string) (model.Resources, error) {
+	resources := model.Resources{}
+	for _, resourceName := range resourceNames {
+		secret, err := s.st.GenerateSecret(resourceName)
 		if err != nil {
-			sdsServiceLog.Errorf("Failed to get credential token through credential fetcher: %v", err)
-			return "", err
+			// Typically, in Istiod, we do not return an error for a failure to generate a resource
+			// However, here it makes sense, because we are generally streaming a single resource,
+			// so sending an error will not cause a single failure to prevent the entire multiplex stream
+			// of resources, and failures here are generally due to temporary networking issues to the CA
+			// rather than a result of configuration issues, which trigger updates in Istiod when resolved.
+			// Instead, we rely on the client to retry (with backoff) on failures.
+			return nil, fmt.Errorf("failed to generate secret for %v: %v", resourceName, err)
 		}
-		token = t
-	} else {
-		tok, err := ioutil.ReadFile(s.jwtPath)
-		if err != nil {
-			sdsServiceLog.Errorf("failed to get credential token: %v from path %s", err, s.jwtPath)
-			return "", err
-		}
-		token = string(tok)
-	}
-	return token, nil
-}
-
-func (s *sdsservice) Stop() {
-	s.closing <- true
-}
 
-func (s *sdsservice) clearStaledClientsJob() {
-	s.ticker = time.NewTicker(s.tickerInterval)
-	for {
-		select {
-		case <-s.ticker.C:
-			clearStaledClients()
-		case <-s.closing:
-			if s.ticker != nil {
-				s.ticker.Stop()
-			}
-		}
+		res := util.MessageToAny(toEnvoySecret(secret))
+		resources = append(resources, res)
 	}
+	return resources, nil
 }
 
-func clearStaledClients() {
-	sdsServiceLog.Debug("start staled connection cleanup job")
-	sdsClientsMutex.Lock()
-	defer sdsClientsMutex.Unlock()
-
-	for connKey := range staledClientKeys {
-		sdsServiceLog.Debugf("remove staled clients %+v", connKey)
-		delete(sdsClients, connKey)
-		delete(staledClientKeys, connKey)
-		// totalStaleConnCounts should be 0 when the for loop finishes.
-		totalStaleConnCounts.Decrement()
+func (s *sdsservice) Generate(_ *model.Proxy, _ *model.PushContext, w *model.WatchedResource, updates *model.PushRequest) (model.Resources, error) {
```

Comment:
Could you please add a brief comment above Generate()? Which component is going to call Generate()?
Another question is, what does updates.Full mean?

Reply:
Added a comment. Basically, the xDS server dispatches the Envoy config generation to different implementators of Generate() based on TypeUrl.

For us, we have only one implementation. But for the main istiod xds, there are a bunch of generators for each type supported.

Reply:
Thanks for the information. And about the second question, by reading the comments in [type PushRequest struct {}](https://github.com/istio/istio/blob/6b967a0e36fa74bb2ffcf84a37b02b8bf5b415b2/pilot/pkg/model/push_context.go#L279), `// Full determines whether a full push is required or not. If false, an incremental update will be sent.` Is `updates.Full` always hardcoded to true in SDS use case?